x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -126,15 +127,32 @@ protected List<Batch<PhysicalPlan>> batches() {
         return rules(true);
     }
 
-    private static class ReplaceAttributeSourceWithDocId extends OptimizerRule<EsSourceExec> {
+    private static class ReplaceSourceAttributes extends OptimizerRule<EsSourceExec> {
 
-        ReplaceAttributeSourceWithDocId() {
+        ReplaceSourceAttributes() {
             super(UP);
         }
 
         @Override
         protected PhysicalPlan rule(EsSourceExec plan) {
-            return new EsQueryExec(plan.source(), plan.index(), plan.indexMode(), plan.query());
+            var docId = new FieldAttribute(plan.source(), EsQueryExec.DOC_ID_FIELD.getName(), EsQueryExec.DOC_ID_FIELD);
+            if (plan.indexMode() == IndexMode.TIME_SERIES) {
+                Attribute tsid = null, timestamp = null;
+                for (Attribute attr : plan.output()) {
+                    String name = attr.name();
+                    if (name.equals(MetadataAttribute.TSID_FIELD)) {
+                        tsid = attr;
+                    } else if (name.equals(MetadataAttribute.TIMESTAMP_FIELD)) {
+                        timestamp = attr;
+                    }
+                }
+                if (tsid == null || timestamp == null) {
```

Comment:
What's the benefit of matching them from plan.output() here, is it to support piped operations propagating `_tsid` and `@timestamp` fields? Otherwise, these are guaranteed to be present in indexes with `tsdb` mode so the current logic populating these in `EsQueryExec` should suffice?

Reply:
The reason is that we use nameId instead of the name (or alias) of expressions to wire layouts during planning.